Release 2.4.0 (#9674)
ZcashFoundation/zebra: CHANGELOG.md

Author: Removed entries:

```
- feat(ci): Enable stateful MIGs and fix release deployment skipping ([#9516](https://github.com/ZcashFoundation/zebra/pull/9516))
- fix link in docs ([#9628](https://github.com/ZcashFoundation/zebra/pull/9628))
- fix(ci): delete scan-test ([#9530](https://github.com/ZcashFoundation/zebra/pull/9530))
- fix(rpc): Use `RawValue` instead of `Value` ([#9474](https://github.com/ZcashFoundation/zebra/pull/9474))
- fix(rpc): reuse txid in getrawtransaction ([#9650](https://github.com/ZcashFoundation/zebra/pull/9650))
- refactor(ci): Simplify cached disk selection logic ([#9545](https://github.com/ZcashFoundation/zebra/pull/9545))
- refactor(test)!: standardize naming conventions across CI and Rust ([#9621](https://github.com/ZcashFoundation/zebra/pull/9621))
- rpc: add deserialization tests ([#9550](https://github.com/ZcashFoundation/zebra/pull/9550))
- tower-batch-control: Fix `Batch::new` compilation with `tokio_unstable` ([#9547](https://github.com/ZcashFoundation/zebra/pull/9547))
- update README with Arch build patch ([#9513](https://github.com/ZcashFoundation/zebra/pull/9513))
- change(deps): Allow Zebra crates to be compiled with alternative versions of their dependencies ([#9484](https://github.com/ZcashFoundation/zebra/pull/9484))
- refactor(ci): Run standard GitHub runners in forks and clones ([#9362](https://github.com/ZcashFoundation/zebra/pull/9362))
- feat(ci): Enable stateful MIGs and fix release deployment skipping ([#9516](https://github.com/ZcashFoundation/zebra/pull/9516))
- fix(ci): handle tests that save state without requiring a cached disk ([#9582](https://github.com/ZcashFoundation/zebra/pull/9582))
- fix(ci): mark `sync-to-checkpoint` test as long-running ([#9595](https://github.com/ZcashFoundation/zebra/pull/9595))
- fix(ci): pin lightwalletd to `v0.4.17` to prevent CI failures ([#9575](https://github.com/ZcashFoundation/zebra/pull/9575))
- fix(ci): prevent Google Cloud workflow failures on fork PRs ([#9573](https://github.com/ZcashFoundation/zebra/pull/9573))
- refactor(ci): Run standard GitHub runners in forks and clones ([#9362](https://github.com/ZcashFoundation/zebra/pull/9362))
- refactor(ci): simplify GCP integration with `create-with-container` ([#9479](https://github.com/ZcashFoundation/zebra/pull/9479))
-
- Update `CONTRIBUTING.md` ([#9561](https://github.com/ZcashFoundation/zebra/pull/9561))
- add(process): Add hotfix release checklist ([#9544](https://github.com/ZcashFoundation/zebra/pull/9544))
- build(deps): bump the devops group across 1 directory with 6 updates ([#9476](https://github.com/ZcashFoundation/zebra/pull/9476))
- build(deps): bump the ecc group across 1 directory with 3 updates ([#9577](https://github.com/ZcashFoundation/zebra/pull/9577))
- build(deps): bump the prod group across 1 directory with 12 updates ([#9651](https://github.com/ZcashFoundation/zebra/pull/9651))
- fix(ci): mark `sync-to-checkpoint` test as long-running ([#9595](https://github.com/ZcashFoundation/zebra/pull/9595))
- fix(clippy, build): fix latest Clippy lints introduced in v1.87.0 and migrate from vergen to vergen\_git ([#9531](https://github.com/ZcashFoundation/zebra/pull/9531))
- fix(rpc): Regroup imports ([#9521](https://github.com/ZcashFoundation/zebra/pull/9521))
- refactor(ci): Run standard GitHub runners in forks and clones ([#9362](https://github.com/ZcashFoundation/zebra/pull/9362))
```

Reviewer: Should we include the doc/README update in https://github.com/ZcashFoundation/zebra/pull/9513 somewhere in the changelog?

Author: Added in 293201e04

### Breaking Changes

This release has the following breaking changes:
- *TODO*: Check the `Removed` and `Deprecated` sections for any breaking changes
- *TODO*: Add a short description of the user impact of each breaking change, and any actions users need to take

- This release contains a major database upgrade. It will upgrade an existing
  database, automatically moving it from the `v26/` folder to a new `v27/`
  folder. However, downgrading is not possible. If you want to keep the
  possibility of reverting Zebra in case of an unexpected issue, backup the

>   `v26/` folder _before_ running the upgraded Zebra. Note that the upgrade takes
>   around 4~6 hours to complete depending on the machine capabilites. Zebra will

Reviewer: ```suggestion
  `v26/` folder _before_ running the upgraded Zebra. Note that the upgrade is slow
  and could take several hours or more to complete on smaller machines. Zebra will
```

  operate normally during that time; the only difference is that some RPC
  responses might return empty or not accurate data (pool values for arbitrary
  block heights and received balances for addresses).
- The `debug_like_zcashd` config option for mining is no longer available. It
  was not enabled by default; if it is now present in the config file, Zebra
  will panic. Simply delete the config option to fix.

### Security

- fix: Pad script in V5 coinbase ([#9620](https://github.com/ZcashFoundation/zebra/pull/9620))

### Added

> - Transaction and block deserialization ([#9522](https://github.com/ZcashFoundation/zebra/pull/9522))

Reviewer: ```suggestion
- Implemented deserialization for Zebra's block and transaction types ([#9522](https://github.com/ZcashFoundation/zebra/pull/9522))
```

> - tests(wallet): Add initial framework to test zallet RPC methods ([#9554](https://github.com/ZcashFoundation/zebra/pull/9554))

Reviewer: This one doesn't seem user-visible.

```suggestion
```

> - change(rpc): Update `getaddressbalance` RPC to return `received` field ([#9295](https://github.com/ZcashFoundation/zebra/pull/9295))

Reviewer: ```suggestion
- Update `getaddressbalance` RPC to return `received` field ([#9295](https://github.com/ZcashFoundation/zebra/pull/9295), [#9539](https://github.com/ZcashFoundation/zebra/pull/9539))
```

> - add(rpc): Add a `mempool_change()` RPC method for listening to changes in the mempool ([#9494](https://github.com/ZcashFoundation/zebra/pull/9494))

Reviewer: Nitpick:

```suggestion
- Added a `mempool_change()` gRPC method for listening to changes in the mempool ([#9494](https://github.com/ZcashFoundation/zebra/pull/9494))
```

- Add `raw_value` feature to serde\_json ([#9538](https://github.com/ZcashFoundation/zebra/pull/9538))

> - feat(config): Modify config to use IPv6 listen\_addr by default ([#9609](https://github.com/ZcashFoundation/zebra/pull/9609))

Reviewer: Nitpick/Subjective:

```suggestion
- Modify `zebra_network::Config` type to use IPv6 listen\_addr by default ([#9609](https://github.com/ZcashFoundation/zebra/pull/9609))
```

- feat(rpc): Add `invalidateblock` and `reconsiderblock` RPC methods ([#9551](https://github.com/ZcashFoundation/zebra/pull/9551))
- feat(rpc): Update `(z_)validateaddress` to validate TEX addresses ([#9483](https://github.com/ZcashFoundation/zebra/pull/9483))

> - feat(rpc): `addnode` method ([#9604](https://github.com/ZcashFoundation/zebra/pull/9604))

Reviewer: ```suggestion
- Added a `addnode` RPC method ([#9604](https://github.com/ZcashFoundation/zebra/pull/9604))
```

- feat(rpc): add missing fields to getrawtransaction ([#9636](https://github.com/ZcashFoundation/zebra/pull/9636))

> - feat(rpc): add value pools to getblock ([#9432](https://github.com/ZcashFoundation/zebra/pull/9432))

Reviewer: Do we want to reference the db format upgrade here as well?

```suggestion
- Added value pool balances to `getblock` RPC output ([#9432](https://github.com/ZcashFoundation/zebra/pull/9432), [#9539](https://github.com/ZcashFoundation/zebra/pull/9539))
```

- add: Support configuring shielded addresses for mining ([#9574](https://github.com/ZcashFoundation/zebra/pull/9574))

> ### Changed

Reviewer: Do we want to mention https://github.com/ZcashFoundation/zebra/pull/9484 as a change here as well?

Author: Added in 293201e04

> - rpc: API cleanup ([#9568](https://github.com/ZcashFoundation/zebra/pull/9568))

Reviewer: ```suggestion
- Renamed and moved exports in `zebra-rpc` ([#9568](https://github.com/ZcashFoundation/zebra/pull/9568))
```

- change(state): Upgrade db format to support new fields in RPC outputs ([#9539](https://github.com/ZcashFoundation/zebra/pull/9539))
- change(rpc): Move GBT RPCs into the main RPC server ([#9459](https://github.com/ZcashFoundation/zebra/pull/9459))
- change(consensus): Adds a `Nu6_1` variant to `NetworkUpgrade` ([#9526](https://github.com/ZcashFoundation/zebra/pull/9526))
- Use zcash\_script’s new `Script` trait ([#8751](https://github.com/ZcashFoundation/zebra/pull/8751))
- Remove `debug_like_zcashd` config option ([#9627](https://github.com/ZcashFoundation/zebra/pull/9627))

> ### Fixed

Reviewer: Do we want to mention https://github.com/ZcashFoundation/zebra/pull/9667 and https://github.com/ZcashFoundation/zebra/pull/9474 (as ~"preserve order of RPC output fields") here as well?

Author: Good idea, added in 293201e04

> - fix(consensus): reuse sighasher when validating txs ([#9594](https://github.com/ZcashFoundation/zebra/pull/9594))

Reviewer: Nitpick/Subjective:

```suggestion
- fix(consensus): Avoid repeatedly converting transactions to librustzcash types when computing sighashes ([#9594](https://github.com/ZcashFoundation/zebra/pull/9594))
```

> - zebra-rpc: Correctly set optional `scriptPubKey` fields of transactions ([#9536](https://github.com/ZcashFoundation/zebra/pull/9536))

Reviewer: Nitpick/Subjective:

```suggestion
- Correctly set optional `scriptPubKey` fields of transactions in `getblock` and `getrawtransaction` RPC outputs ([#9536](https://github.com/ZcashFoundation/zebra/pull/9536))
```

- fix(network): Allow local outbound connections on Regtest ([#9580](https://github.com/ZcashFoundation/zebra/pull/9580))

> - fix(rpc): `z_gettreestate` optional fields ([#9451](https://github.com/ZcashFoundation/zebra/pull/9451))

Reviewer: ```suggestion
- Allow for parsing `z_gettreestate` output type where optional fields are omitted ([#9451](https://github.com/ZcashFoundation/zebra/pull/9451))
```

- fix: Pad script in V5 coinbase ([#9620](https://github.com/ZcashFoundation/zebra/pull/9620))

### Contributors

Thank you to everyone who contributed to this release, we couldn't make Zebra without you:

> @ala-mode, @arya2, @conradoplg, @dependabot[bot], @elijahhampton, @gustavovalverde, @idky137, @mpguerra, @oxarbitrage, @sellout, @str4d and @upbqdn

Reviewer: ```suggestion
@ala-mode, @arya2, @conradoplg, @elijahhampton, @gustavovalverde, @idky137, @mpguerra, @oxarbitrage, @sellout, @str4d and @upbqdn
```

- This release contains a major database upgrade. It will upgrade an existing
  database, automatically moving it from the `v26/` folder to a new `v27/`
  folder. However, downgrading is not possible. If you want to keep the
  possibility of reverting Zebra in case of an unexpected issue, backup the
  `v26/` folder _before_ running the upgraded Zebra. Note that the upgrade is slow
  and could take several hours or more to complete on smaller machines. Zebra will
  operate normally during that time; the only difference is that some RPC
  responses might return empty or not accurate data (pool values for arbitrary
  block heights and received balances for addresses).
- The `debug_like_zcashd` config option for mining is no longer available. It
  was not enabled by default; if it is now present in the config file, Zebra
  will panic. Simply delete the config option to fix.
- The `cm_u` field byte order was fixed in `getrawtransaction`/`getblock`
  response, so if you relied on the wrong order, you will need to fix your
  application.
- The `zebra-scan` and `zebra-grpc` crates are no longer supported and were
  removed from the codebase.

### Security

> - fix: Pad script in V5 coinbase ([#9620](https://github.com/ZcashFoundation/zebra/pull/9620))

Reviewer 2: ```suggestion
- Fix a consensus rule violation in V5 coinbase transactions at low heights. This issue could only occur on Regtest or custom testnets and is now resolved  ([#9620](https://github.com/ZcashFoundation/zebra/pull/9620))
```

- Modified `zebra_network::Config` type to use IPv6 listen\_addr by default ([#9609](https://github.com/ZcashFoundation/zebra/pull/9609))
- Added `invalidateblock` and `reconsiderblock` RPC methods ([#9551](https://github.com/ZcashFoundation/zebra/pull/9551))
- Updated `(z_)validateaddress` to validate TEX addresses ([#9483](https://github.com/ZcashFoundation/zebra/pull/9483))
- Added a `addnode` RPC method ([#9604](https://github.com/ZcashFoundation/zebra/pull/9604))
- Added missing fields to getrawtransaction ([#9636](https://github.com/ZcashFoundation/zebra/pull/9636))
- Added value pool balances to `getblock` RPC output ([#9432](https://github.com/ZcashFoundation/zebra/pull/9432), [#9539](https://github.com/ZcashFoundation/zebra/pull/9539))
- Added support for configuring shielded addresses for mining ([#9574](https://github.com/ZcashFoundation/zebra/pull/9574))
- Added binding_sig, joinsplit_pub_key and joinsplit_sig fields to `getrawtransaction`/`getblock` response ([#9652](https://github.com/ZcashFoundation/zebra/pull/9652))
- Added a method in `zebra-rpc` to allow validating addresses ([#9658](https://github.com/ZcashFoundation/zebra/pull/9658))

> - Added support for generating V4 coinbase transactions ([#9647](https://github.com/ZcashFoundation/zebra/pull/9647))

Reviewer 2: ```suggestion
```

This was already possible with the `debug_like_zcashd` option, we lost support for a while in `main` but never in release. We can just remove it from the list.

### Changed

- Allow Zebra crates to be compiled with alternative versions of their dependencies ([#9484](https://github.com/ZcashFoundation/zebra/pull/9484))
- Updated README with Arch build patch ([#9513](https://github.com/ZcashFoundation/zebra/pull/9513))
- Renamed and moved exports in `zebra-rpc` ([#9568](https://github.com/ZcashFoundation/zebra/pull/9568))
- Upgraded DB format to support new fields in RPC outputs ([#9539](https://github.com/ZcashFoundation/zebra/pull/9539))
- Moved GBT RPCs into the main RPC server ([#9459](https://github.com/ZcashFoundation/zebra/pull/9459))
- Added a `Nu6_1` variant to `NetworkUpgrade` ([#9526](https://github.com/ZcashFoundation/zebra/pull/9526))
- Use zcash\_script’s new `Script` trait ([#8751](https://github.com/ZcashFoundation/zebra/pull/8751))
- Removed `debug_like_zcashd` config option ([#9627](https://github.com/ZcashFoundation/zebra/pull/9627))

> - Sync all chains in TrustedChainSync::sync, add `NonFinalizedStateChange` gRPC method ([#9654](https://github.com/ZcashFoundation/zebra/pull/9654))

Reviewer 2: ```suggestion
- Sync all chains in `TrustedChainSync::sync`, add `NonFinalizedStateChange` gRPC method ([#9654](https://github.com/ZcashFoundation/zebra/pull/9654))
```

- Added `prometheus` as a default feature in zebrad ([#9677](https://github.com/ZcashFoundation/zebra/pull/9677))

### Fixed

- Preserve order of RPC output fields ([#9474](https://github.com/ZcashFoundation/zebra/pull/9474))
- Fixed `cm_u` field byte order in `getrawtransaction`/`getblock` response ([#9667](https://github.com/ZcashFoundation/zebra/pull/9667))
- Avoid repeatedly converting transactions to librustzcash types when computing sighashes ([#9594](https://github.com/ZcashFoundation/zebra/pull/9594))
- Correctly set optional `scriptPubKey` fields of transactions in `getblock` and `getrawtransaction` RPC outputs ([#9536](https://github.com/ZcashFoundation/zebra/pull/9536))
- Allow local outbound connections on Regtest ([#9580](https://github.com/ZcashFoundation/zebra/pull/9580))
- Allow for parsing `z_gettreestate` output type where optional fields are omitted ([#9451](https://github.com/ZcashFoundation/zebra/pull/9451))

> - Pad script in V5 coinbase ([#9620](https://github.com/ZcashFoundation/zebra/pull/9620))

Reviewer 2: ```suggestion
```

Repeated.
